cli/tools/pm/audit.rs

```diff
@@ -0,0 +1,538 @@
+
+  use super::*;
+
+  #[derive(Clone, Copy, Debug, PartialEq, PartialOrd)]
+  enum AdvisorySeverity {
+    Low,
+    Moderate,
+    High,
+    Critical,
+  }
+
+  impl AdvisorySeverity {
+    fn parse(str_: &str) -> Option<Self> {
+      match str_ {
+        "low" => Some(Self::Low),
+        "moderate" => Some(Self::Moderate),
+        "high" => Some(Self::High),
+        "critical" => Some(Self::Critical),
+        _ => None,
+      }
+    }
+  }
+
+  fn get_dependency_descriptors_for_deps(
+    seen: &mut HashSet<String>,
+    npm_resolution_snapshot: &NpmResolutionSnapshot,
+    package_id: &NpmPackageId,
+  ) -> HashMap<String, Box<DependencyDescriptor>> {
+    let resolution_package =
+      npm_resolution_snapshot.package_from_id(package_id).unwrap();
+    let mut deps_map =
+      HashMap::with_capacity(resolution_package.dependencies.len());
+    for dep in resolution_package.dependencies.iter() {
+      let seen_str =
+        format!("{}@{}", dep.0.to_string(), dep.1.nv.version.to_string());
+      if seen.contains(&seen_str) {
+        continue;
+      }
+      seen.insert(seen_str);
+
+      let dep_deps = get_dependency_descriptors_for_deps(
+        seen,
+        npm_resolution_snapshot,
+        dep.1,
+      );
+      deps_map.insert(
+        dep.0.to_string(),
+        Box::new(DependencyDescriptor {
+          version: dep.1.nv.version.to_string(),
+          // TODO(bartlomieju): not sure how to determine that from the snapshot
+          dev: false,
+          requires: dep_deps
+            .iter()
+            .map(|(k, v)| (k.to_string(), v.version.to_string()))
+            .collect(),
+          dependencies: dep_deps,
+        }),
+      );
+    }
+    deps_map
+  }
+
+  pub async fn call_audits_api_inner(
+    npm_url: Url,
+    client: HttpClient,
+    body: serde_json::Value,
+  ) -> Result<AuditResponse, AnyError> {
+    let url = npm_url.join("/-/npm/v1/security/audits").unwrap();
+    let future = client.post_json(url, &body)?.send().boxed_local();
+    let response = future.await?;
+    let json_str = http_util::body_to_string(response)
+      .await
+      .context("Failed to read response from the npm registry API")?;
+    let response: AuditResponse = serde_json::from_str(&json_str)
+      .context("Failed to deserialize response from the npm registry API")?;
+    Ok(response)
+  }
+
+  pub async fn call_audits_api(
+    audit_flags: AuditFlags,
+    npm_url: Url,
+    npm_resolution_snapshot: &NpmResolutionSnapshot,
+    client: HttpClient,
+  ) -> Result<i32, AnyError> {
+    let top_level_packages = npm_resolution_snapshot.top_level_packages();
+    let mut requires = HashMap::with_capacity(top_level_packages.len());
+    let mut dependencies = HashMap::with_capacity(top_level_packages.len());
+    let seen = &mut HashSet::with_capacity(top_level_packages.len() * 100);
+    for package in top_level_packages {
+      requires
+        .insert(package.nv.name.to_string(), package.nv.version.to_string());
```

Comment:
It's possible for there to be multiple package versions with the same name. Seems like this will overwrite in some cases and miss advisories?